spark-operator-api/src/main/java/org/apache/spark/k8s/operator/status/BaseStatus.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.k8s.operator.status;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.TreeMap;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonInclude;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.ToString;
+
+@ToString
+@EqualsAndHashCode
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class BaseStatus<S, STATE extends BaseState<S>, AS extends BaseAttemptSummary> {
+  @Getter STATE currentState;
+  @Getter Map<Long, STATE> stateTransitionHistory;
+  @Getter AS previousAttemptSummary;
+  @Getter AS currentAttemptSummary;
+
+  public BaseStatus(STATE initState, AS currentAttemptSummary) {
+    this.currentState = initState;
+    this.stateTransitionHistory = new TreeMap<>();
+    this.stateTransitionHistory.put(0L, initState);
+    this.previousAttemptSummary = null;
+    this.currentAttemptSummary = currentAttemptSummary;
+  }
+
+  public BaseStatus(
+      STATE currentState,
+      Map<Long, STATE> stateTransitionHistory,
+      AS previousAttemptSummary,
+      AS currentAttemptSummary) {
+    this.currentState = currentState;
+    this.stateTransitionHistory = new TreeMap<>(stateTransitionHistory);
+    this.previousAttemptSummary = previousAttemptSummary;
+    this.currentAttemptSummary = currentAttemptSummary;
+  }
+
+  protected long getCurrentStateId() {
+    return Collections.max(stateTransitionHistory.keySet());
```

Comment:
May I ask why we need to use this inefficient way instead of the built-in `lastKey` Java method?

FYI, `lastKey` exists for all `SortedMap`-derived java class (Maybe since Java 1.2?).
- https://docs.oracle.com/javase/8/docs/api/java/util/TreeMap.html#lastKey